openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -781,149 +741,143 @@ def create_library_block(library_key, block_type, definition_id):
     """
-    Create a new XBlock definition in this library of the specified type (e.g.
-    "html"), and add it as a child of the specified existing block.
+    Create a Component for an XBlock type, and initialize it.
+
+    This will create a Component, along with its first ComponentVersion. The tag
+    in the OLX will have no attributes, e.g. `<problem />`. This first version
+    will be set as the current draft. This function does not publish the
+    Component.
 
-    The 'definition_id' value (which should be a string like "problem1") will be
-    used as both the definition_id and the usage_id of the child.
+    TODO: We should probably shift this to openedx.core.djangoapps.xblock.api
+    (along with its caller) since it gives runtime storage specifics. The
+    Library-specific logic stays in this module, so "create a block for my lib"
+    should stay here, but "making a block means creating a component with
+    text data like X" goes in xblock.api.
     """
-    assert isinstance(parent_usage_key, LibraryUsageLocatorV2)
-    # Load the parent block to make sure it exists and so we can modify its 'children' field:
-    parent_block = load_block(parent_usage_key, user=None)
-    if not parent_block.has_children:
-        raise ValueError("The specified parent XBlock does not allow child XBlocks.")
-    # Create the new block in the library:
-    metadata = create_library_block(parent_usage_key.context_key, block_type, definition_id)
-    # Set the block as a child.
-    # This will effectively "move" the newly created block from being a top-level block in the library to a child.
-    include_data = XBlockInclude(link_id=None, block_type=block_type, definition_id=definition_id, usage_hint=None)
-    parent_block.runtime.add_child_include(parent_block, include_data)
-    parent_block.save()
-    ref = ContentLibrary.objects.get_by_key(parent_usage_key.context_key)  # type: ignore[attr-defined]
-    LIBRARY_BLOCK_UPDATED.send_event(
+    display_name = xblock_type_display_name(usage_key.block_type)
+    now = datetime.now(tz=timezone.utc)
+    xml_text = f'<{usage_key.block_type} />'
+
+    learning_package = content_lib.learning_package
+
+    component_type = components_api.get_or_create_component_type(
+        "xblock.v1", usage_key.block_type
+    )
+    component, component_version = components_api.create_component_and_version(
+        learning_package.id,
+        component_type=component_type,
+        local_key=usage_key.block_id,
+        title=display_name,
+        created=now,
+        created_by=None,
+    )
+    content = contents_api.get_or_create_text_content(
+        learning_package.id,
+        get_or_create_olx_media_type(usage_key.block_type).id,
+        text=xml_text,
+        created=now,
+    )
+    components_api.create_component_version_content(
+        component_version.pk,
+        content.id,
+        key="block.xml",
+        learner_downloadable=False
+    )
```

Comment:
@kdmccormick: My comment about whether this should be in an atomic block is in reference to this part, where we're making a bunch of separate objects for components, component versions, and content here, and this should probably be wrapped together in an `with transaction.atomic():`

Reply:
Wrapped